test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/DocumentGenerator.java

```diff
@@ -0,0 +1,89 @@
+
+import org.elasticsearch.logsdb.datageneration.datasource.DataSourceRequest;
+import org.elasticsearch.logsdb.datageneration.datasource.DataSourceResponse;
+
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.Optional;
+import java.util.TreeMap;
+
+public class DocumentGenerator {
+    private final DataGeneratorSpecification specification;
+
+    private final DataSourceResponse.ObjectArrayGenerator objectArrayGenerator;
+
+    public DocumentGenerator(DataGeneratorSpecification specification) {
+        this.specification = specification;
+
+        this.objectArrayGenerator = specification.dataSource().get(new DataSourceRequest.ObjectArrayGenerator());
+    }
+
+    public Map<String, Object> generate(Template template, Mapping mapping) {
+        var documentMap = new TreeMap<String, Object>();
+        for (var predefinedField : specification.predefinedFields()) {
+            documentMap.put(predefinedField.name(), predefinedField.generator(specification.dataSource()).generateValue());
+        }
+
+        generateDocument(documentMap, template.template(), new Context("", mapping.lookup()));
+        return documentMap;
+    }
+
+    private void generateDocument(Map<String, Object> document, Map<String, Template.Entry> template, Context context) {
+        for (var entry : template.entrySet()) {
+            String fieldName = entry.getKey();
+            Template.Entry templateEntry = entry.getValue();
+
+            if (templateEntry instanceof Template.Leaf leaf) {
+                // Unsigned long does not play well when dynamically mapped because
+                // it gets mapped as just long and large values fail to index.
+                // Just skip it.
+                if (leaf.type() == FieldType.UNSIGNED_LONG && context.mappingLookup().get(context.pathTo(fieldName)) == null) {
+                    continue;
+                }
+
+                var generator = leaf.type().generator(fieldName, specification.dataSource());
+
+                document.put(fieldName, generator.generateValue());
+            } else if (templateEntry instanceof Template.Object object) {
+                Optional<Integer> arrayLength = objectArrayGenerator.lengthGenerator().get();
+
+                if (arrayLength.isPresent()) {
+                    var children = new ArrayList<>(arrayLength.get());
+                    document.put(object.name(), children);
+
+                    for (int i = 0; i < arrayLength.get(); i++) {
+                        children.add(generateObject(object, context));
+                    }
+                } else {
+                    document.put(object.name(), generateObject(object, context));
+                }
+            }
+        }
+    }
+
+    private Map<String, Object> generateObject(Template.Object object, Context context) {
+        var children = new TreeMap<String, Object>();
+        generateDocument(children, object.children(), context.stepIntoObject(object.name()));
```

Comment:
Nit: it's somewhat confusing that `generateObject` calls `generateDocument`.. Maybe rename the latter `generateContents` or `generateFields`